### PR TITLE
IO Refactoring: Code Scheme Tests

### PR DESCRIPTION
**Pull request by as2388**
Includes some tests for saving and loading code schemes.

Not all of them are passing - I would prefer @lukechurch to check that the failing tests reflect desired behaviour before fixing anything.

Also includes some documentation to Code, which was necessary to understand how to construct some test CodeSchemes.

### CODE REVIEW

File: coda-prototype/package.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "coda",
+  "version": "1.0.0",
+  "description": "Coda is an interactive tool that helps you label short text datasets!",
+  "main": "ui.js",
+  "dependencies": {
+    "karma": "^1.7.1",
```
Comment (lukechurch) [inline coda-prototype/package.json]: Why are these dependencies not dev dependencies?

File: coda-prototype/src/io/StorageUtils.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
+class StorageManager {
+    constructor() {
+        /*
```
Comment (lukechurch) [inline coda-prototype/src/io/StorageUtils.js]: Why the large commented out code block?

File: coda-prototype/src/io/StorageUtils.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                    }
+
+                }
+            }
+        });*/
+    }
+    static get instance() {
+        return this._instance || (this._instance = new StorageManager());
+    }
+    isExpired() {
```
Comment (lukechurch) [inline coda-prototype/src/io/StorageUtils.js]: Could we move this into TS so we could start putting type annotations on these functions?

File: coda-prototype/src/io/StorageUtils.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        // TODO
```
Comment (lukechurch) [inline coda-prototype/src/io/StorageUtils.js]: This TODO is in the wrong place - this function looks like it does approximately the right thing, so the TODO is to use it?

File: coda-prototype/src/io/StorageUtils.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        // if yes, clear storage
+        var _MS_PER_DAY = 1000 * 60 * 60 * 24;
+        // a and b are javascript Date objects
+        function dateDiffInDays(a, b) {
+            // Discard the time and time-zone information.
+            let utc1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
+            let utc2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());
+            return Math.floor((utc2 - utc1) / _MS_PER_DAY);
+        }
+        return (30 <= dateDiffInDays(this.lastEdit, new Date()));
```
Comment (lukechurch) [inline coda-prototype/src/io/StorageUtils.js]: This logic feels backwards compared to how I would read it "the last edit was more than 30 days ago" - can we reverse the conditional?

File: coda-prototype/src/io/StorageUtils.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                if (typeof dataset == "string") {
+                    dataset = JSON.parse(dataset);
+                }
+                if (data == null || dataset == null || typeof data == 'undefined' ||
+                    typeof dataset == 'undefined' || Object.keys(dataset).length == 0) {
+                    reject(new Error("No valid dataset available in storage!"));
+                }
+                if (!dataset.hasOwnProperty("schemes") || !dataset.hasOwnProperty("sessions") ||
+                    !dataset.hasOwnProperty("events") || Object.keys(dataset["events"]).length == 0) {
+                    reject(new Error("Reading from storage failed - dataset format is corrupt."));
```
Comment (lukechurch) [inline coda-prototype/src/io/StorageUtils.js]: Please log more indication as to what the problem is than this

File: coda-prototype/src/io/StorageUtils.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                            instr = JSON.parse(data["instrumentation"]).concat(activity);
+                        }
+                        else {
+                            instr = activity;
+                        }
+                        chrome.storage.local.set({ "instrumentation": JSON.stringify(instr), }, () => {
+                            if (chrome.runtime.lastError) {
+                                console.log(chrome.runtime.lastError);
+                            }
+                            else {
```
Comment (lukechurch) [inline coda-prototype/src/io/StorageUtils.js]: I find the deep nested if-else blocks very hard to read. Can they be flattened into an easier pattern?

File: coda-prototype/package.json (hunk 1 above)
Comment (as2388) [reply coda-prototype/package.json]: fixed

File: coda-prototype/src/io/StorageUtils.js (hunk 2 above)
Comment (as2388) [reply coda-prototype/src/io/StorageUtils.js]: This file has yet to be refactored; these comments will be addressed in future when we improve storageUtils.

**Comment by as2388** [comment]
@lukechurch Test file with failing tests marked with comments:
https://github.com/lukechurch/coda/blob/eb174b268926722428e7dd37af8f01e4f903f26f/coda-prototype/test/FileUtils.test.ts

**Pull request merged**